### DISCUSSION TITLE
Looking for a solution to include the rotational DoFs into calculs

### DISCUSSION THREAD

**Discussion by ZAARAOUI999** [adtzlr/felupe#309] [Q&A] [answered]
Hi all, I'am asking if is it possible to use the idea described below in order to add the missing rotational DoFs:
So, we have the three known matrix of rotation which defined as:
```

Rx = np.array([[1, 0, 0],
      [0, np.cos(t), -np.sin(t)],
      [0, np.sin(t), np.cos(t)]])

Ry = np.array([[np.cos(t), 0, np.sin(t)],
      [0, 1, 0],
      [-np.sin(t), 0, np.cos(t)]])

Rz = np.array([[np.cos(t), -np.sin(t), 0],
      [np.sin(t), np.cos(t), 0],
      [0, 0, 1]])

```

We can assign (may we can say 'map') to each node (or a group of nodes) a rotation matrix, so the new position of a node is the old position plus the translation plus the rotation.
The rotation angle t will be treated as the bounds['move'] in the ramp dict aka bounds['rotate']=sth like 1D array which describe a rotation, so we may add sth to the 'Boundary' code like (we have fx, fy, fx and we can add rx, ry, rz  or we have skip(Tx, Ty, Tz) and we can add skip(Rx, Ry, Rz) for example). Now, let's take a mesh and rotate a portion of it about the z-axis with 30°:

```
t = np.deg2rad(30)
Rz = np.array([[np.cos(t), -np.sin(t), 0],
          [np.sin(t), np.cos(t), 0],
          [0, 0, 1]])
		  
mesh.points[:500] = np.dot(mesh.points[:500],Rz)

```  

![unrotated](https://user-images.githubusercontent.com/115699524/196496933-1d72db83-b817-4f64-96e7-23a2a8325596.PNG)
![rotated](https://user-images.githubusercontent.com/115699524/196496965-2179e4ec-ab19-4b8a-8369-4cef75e9394f.PNG)

Can this give any idea to solve the missing dofs?

**Comment by adtzlr** [comment, marked as answer]
What about this one: Use the Mesh-helper `rotate` to obtain the displacements on the top boundary due to the enforced rotation and apply the enforced displacements using the `values` argument of a boundary. Alternatively, there is also a helper for the rotation matrix `fem.math.rotation_matrix()`.

```python
import felupe as fem
import numpy as np

# create a hexahedron-region on a cube
mesh = fem.Cube(n=11)
region = fem.RegionHexahedron(mesh)

# add a mixed field container (with displacement, pressure and volume ratio)
field = fem.FieldsMixed(region, n=3, values=(0, 0, 1))

# apply a uniaxial elongation on the cube
boundaries = dict(
    bottom=fem.Boundary(field[0], fz=0),
    top=fem.Boundary(field[0], fz=1),
)

# create the values for the rotation
move = []
for angle_deg in np.linspace(0, 50, 6):
    
    # rotate point coordinates on top boundary
    top = mesh.points[boundaries["top"].points]
    top_rotated = fem.mesh.rotate(
        points=top, 
        cells=None, 
        cell_type=None,
        angle_deg=angle_deg, 
        axis=2, 
        center=[0, 0, 0]
    )[0]

    move.append((top_rotated - top).ravel())

# define the constitutive material behaviour and create a solid body
umat = fem.ThreeFieldVariation(fem.NeoHooke(mu=1, bulk=5000))
solid = fem.SolidBody(umat, field)

# prepare a step with substeps
step = fem.Step(
    items=[solid], 
    ramp={boundaries["top"]: move}, 
    boundaries=boundaries
)

# add the step to a job, evaluate all substeps and create a plot
job = fem.CharacteristicCurve(steps=[step], boundary=boundaries["top"])
job.evaluate(filename="result.xdmf")
```

**Comment by adtzlr** [comment]
![image](https://user-images.githubusercontent.com/5793153/196743488-71e30acc-82d2-4515-8015-7c73b0f251ed.png)

**Reply by ZAARAOUI999** [reply]
Great!!! Thank you,That's what I were looking for. Just one missing thing is that the rotation should be pure, I mean that the top face should just make a rotation without translation (The center of the top face shouldn't change). Is my question clear?

**Reply by ZAARAOUI999** [reply]
Oh, sorry this problem due to the choice of the center of rotation, I didn't pay attention, thank you again.
